### ISSUE TITLE
Unable to delete a book (constraint violation)

### ISSUE THREAD

**Issue by knjk04** [knjk04/book-project#114]
**Describe the bug**

Sometimes, a book cannot be deleted.

**To Reproduce**
Steps to reproduce the behaviour:
1. Go to the 'read' shelf
2. Change Roald's name to "Roald's"
3. Try deleting "Harry Potter and the Chamber of Secrets"

**Expected behaviour**

The book should be deleted.

**Additional context**

Stack trace:

https://gist.github.com/knjk04/86d7d7406a227ef4bf050e9bf58b3e44

**Comment by judeibe** [comment]
I have not written a solution but I diagnosed the  problem.
The issue is in how the data is pre-populated. The more than one book is added that references the same author. Those books are saved using the same author id. When you try the delete any book without first changing the author (creating a new author reference for that book), the delete command will cascade and try to delete the author as well but will fail because the author is referenced by other books.

`
Caused by: org.h2.jdbc.JdbcSQLIntegrityConstraintViolationException:Referential integrity constraint violation: "FKKLNRV3WELER2FTKWEEWLKY958: PUBLIC.BOOK FOREIGN KEY(AUTHOR_ID) REFERENCES PUBLIC.AUTHOR(ID) (6)"; SQL statement:
delete from author where id=? [23503-200]
`

**Comment by knjk04** [comment]
Thanks a lot for this!

**Comment by knjk04** [comment]
Note to the person who picks this issue up:

One of the problems with this bug is that when a new book is added, a new author is created (even if an author with the same name exists). However, when we populate the repositories in `PredefinedShelfService`, a book may be assigned to the same author.

If the user wants to add a book for an author 'Joe Bloggs', but 'Joe Bloggs' already exists as an author, the user should be able to choose whether to assign this 'Joe Bloggs' as a new author or assign this book to the existing 'Joe Bloggs'.

**Comment by amit1307** [comment]
Hello Karan, I can give this one a try?

**Comment by knjk04** [comment]
Thanks for your interest in this! Before I assign you to the issue, could you explain your implementation approach?

**Comment by amit1307** [comment]
Sure, I haven't thought about the solution yet. I'll get back as soon as I have something in mind.

**Comment by amit1307** [comment]
Hey, had a quick look at it and how about the below:

Currently we have a uni directional mapping between Book and Author with Author as parent. The main issue here is when a child (Book) is deleted the parent (Author) could be referred by another child or children so the delete fails. So the solution to this problem could be:

Create bi-directional mapping between Book and Author with `orphanRemoval = true`. So this way it should be fine. I had a quick test locally and seems fine. What do you think?

Regarding the UI change where if the Author with the same name already exists, I am not sure about this change because:
1. Currently we have only first name and last name for Author so how would having more than 1 Author with same FN and LN make sense for the app?
2. I would say if the Author with the same FN and LN exists we must associate the new Book being added to that Author only and shouldn't allow a new Author with same name to be added. What's the main issue we see with this approach? 

But if you have a strong opinion about the approach you suggested, I am happy to make the change but just trying to understand the thinking behind the change requested.

**Comment by knjk04** [comment]
Thanks for this. The backend change sounds good. There's currently a pull request open that is adding books to the `Author` class (#189 ). so perhaps you could add `orphanRemoval = true` once that's merged in.

1 & 2. There may be 2 authors with the same first and last name, as you pointed out in point 2. These authors may have written different books.

To keep things simple for now, point 2 is the better approach. However, in the future, there may come a time when we need at least two authors with the same with different books. Perhaps the solution for the UI then would be to first make users add an author before adding a book. Then, they have to use an existing author when setting the author in the book form (instead of typing out the first and last name).

I was hoping we could come up with a better solution that what Goodreads currently has for this problem, which seems to be just adding spaces at the end of one of the author's last name (see https://www.goodreads.com/topic/show/1623295-two-authors-with-the-same-name)

However, this UI issue can be a separate issue. For now, this issue can just tackle the backend change needed.

**Comment by knjk04** [comment]
Blocked by issue #110

**Comment by knjk04** [comment]
This is now unblocked a set of books has been added to the `Author` entity

**Comment by amit1307** [comment]
Thanks for letting me know, I'll have a look at it soon.

**Comment by knjk04** [comment]
@amit1307 Do you know when you will be able to take a look?

**Comment by amit1307** [comment]
@knjk04 I have started on it and should raise the PR soon (max by the weekend), hope that's fine.

**Comment by knjk04** [comment]
No problem, just wanted to check that everything was OK.

**Comment by knjk04** [comment]
@amit1307 Do you have another update on this? I'm just asking as this is a high-priority issue and it's blocking an assigned issue for someone else

**Comment by amit1307** [comment]
Hey, I am really sorry for the delay. I was out of town so completely forgot about it. I shall raise the PR tonight/tomorrow morning.

**Comment by knjk04** [comment]
No matter
